src/State/ForValues.lua

```diff
@@ -0,0 +1,325 @@
+					end
+
+					-- remove stored value data
+					meta[outputValue] = nil
+				end
+
+				-- step 2: swap the old/new dependencies; clean out the new dependency set so we get a fresh start
+				valueData.oldDependencySet, valueData.dependencySet =
+					valueData.dependencySet, valueData.oldDependencySet
+				table.clear(valueData.dependencySet)
+
+				-- step 3: remove stored cached value
+				oldValueCache[inValue] = nil
+				cachedValue = nil
+			else
+				-- inform future runs that we already checked this inValue's dependencies,
+				-- and that we don't need to re-capture dependencies
+				dependenciesCaptured[inValue] = true
+			end
+		end
+
+		-- if we still don't need to recalculate, then we can re-use a cached value
+		-- if there are no cached values, then set recalculate to true and
+		-- calculate a new value
+		if not shouldRecalculate then
+			-- if we're using a table, then try to pull a value from the cache
+			if type(cachedValue) == "table" then
+				local cachedValues = cachedValue
+				local cachedValuesSize = #cachedValues
+
+				if cachedValuesSize > 0 then
+					cachedValue = cachedValues[cachedValuesSize]
+					table.remove(cachedValues, cachedValuesSize)
+				else
+					cachedValue = nil
+					shouldRecalculate = true
+				end
+			end
+		end
+
+		-- if we should recalculate the output by this point, do that
+		if shouldRecalculate then
+			local processOK, newOutValue, newMetaValue
+
+			-- if we already captured the dependencies for this inValue, we don't need to do it again
+			if dependenciesCaptured[inValue] then
+				processOK, newOutValue, newMetaValue = pcall(self._processor, inValue)
+
+				-- otherwise, we need to capture the dependencies
+			else
+				processOK, newOutValue, newMetaValue = captureDependencies(
+					valueData.dependencySet,
+					self._processor,
+					inValue
+				)
+			end
+
+			if processOK then
+				-- prepare the value to be cached
+				cachedValue = newOutValue
+				-- store meta value, since we don't touch that when reusing values
+				meta[newOutValue] = newMetaValue
+				-- inform future runs that we already captured the dependencies
+				dependenciesCaptured[inValue] = true
+
+				didChange = true
+			else
+				logErrorNonFatal("forValuesProcessorError", newOutValue)
+			end
+		end
+
+		-- if we successfully created a new value or found a value to reuse,
+		-- cache it and update the stored data
+		if cachedValue ~= nil then
+			-- we store tables and objects in an array of cached objects, since they need to be unique
+			if type(cachedValue) == "userdata" or type(cachedValue) == "table" then
+				local cachedValues = newValueCache[inValue]
+				if not cachedValues then
+					cachedValues = {}
+					newValueCache[inValue] = cachedValues
+				end
+
+				table.insert(cachedValues, cachedValue)
+			else
+				newValueCache[inValue] = cachedValue
+			end
+
+			-- store the value in the output with the same key
+			outputValues[_key] = cachedValue
```

Comment:
The only case cachedValue is nil is when the processor errored, just above. In this case, we're saying the error is nonFatal, but it's leading to the key-value pair silently not being updated. Is this really good behaviour?

Reply:
Yeah I wasn't too sure about this one. We don't currently throw in this scenario in ``ComputedPairs`` so I just followed that. That being said, since we decided to throw on key collisions it also makes sense for this to throw an error. Perhaps @Elttob can explain why we don't currently throw when this happens in ``ComputedPairs``? Does the same logic apply here?

Reply:
I did some thinking on this one. I think this should remain as a ``nonFatal`` error. I don't want to stop all remaining UI updates just because one of your components threw an error during recalculation. 

Also it's not like we're silently not updating it, an error is displayed in the output with a full stack trace.